src/unitxt/inference.py

```diff
@@ -164,86 +175,629 @@ def _is_loaded(self):
+            self.get_return_object(output[0], instance["source"], return_meta_data)
+            if isinstance(output, list)
+            else self.get_return_object(output, instance["source"], return_meta_data)
+            for output, instance in zip(outputs, dataset)
+        ]
+
+    def get_return_object(self, output, inp, return_meta_data):
+        if return_meta_data:
+            return TextGenerationInferenceOutput(
+                prediction=output["generated_text"],
+                model_name=self.model_name,
+                inference_type=self.label,
+                input_text=inp,
+            )
+        return output["generated_text"]
+
+
+class HFPeftInferenceEngine(HFPipelineBasedInferenceEngine):
+    label: str = "hf_peft_pipeline_inference_engine"
+
+    _requirements_list = {
+        "transformers": "Install huggingface package using 'pip install --upgrade transformers",
+        "torch": "Install torch, go on PyTorch website for mode details.",
+        "accelerate": "pip install accelerate",
+        "peft": "Install 'peft' package using: 'pip install peft'.",
+    }
+
+    @staticmethod
+    def is_peft_model(model_name: str) -> bool:
+        from peft import PeftConfig
+
+        try:
+            PeftConfig.from_pretrained(model_name)
+            return True
+        except ValueError as e:
+            if "Can't find" in str(e):
+                return False
+            raise e
+
     def prepare_engine(self):
-        if not self.lazy_load:
-            self._prepare_pipeline()
+        if not self.is_peft_model(self.model_name):
+            raise ValueError(
+                f"Specified model '{self.model_name}' is not the PEFT model. "
+                f"Use a regular instance of the `HFPipelineBasedInferenceEngine` "
+                f"instead."
+            )
+        super().prepare_engine()
+
+    def _define_task(self):
+        from peft import AutoPeftModelForCausalLM, AutoPeftModelForSeq2SeqLM, PeftConfig
+        from transformers import AutoTokenizer
+
+        peft_config = PeftConfig.from_pretrained(self.model_name)
+        assert isinstance(peft_config.base_model_name_or_path, str)
+        original_model_name = self.model_name
+        self.model_name = peft_config.base_model_name_or_path
+        super()._define_task()
+
+        if self.task == "text2text-generation":
+            self.model_name = AutoPeftModelForSeq2SeqLM.from_pretrained(
+                original_model_name
+            )
+        else:
+            self.model_name = AutoPeftModelForCausalLM.from_pretrained(
+                original_model_name
+            )
+
+        self.processor = AutoTokenizer.from_pretrained(
+            peft_config.base_model_name_or_path
+        )
+
+    def _get_model_args(self) -> Dict[str, Any]:
+        args = super()._get_model_args()
+        args["tokenizer"] = self.processor
+        return args
+
+
+class HFInferenceEngine(HFInferenceEngineBase, LogProbInferenceEngine):
```

Comment:
I don't think we need another abstract class. So many is confusing. If this is the log prob class, the name should reflect it and it should be fully implemented. If there are problem with interfaces mismatch, let solves it @elronbandel

Reply:
I wanted to somehow separate pipeline-based HF inference engines (`HFPipelineBasedInferenceEngine` and `HFPeftInferenceEngine`) from "regular" HF inference engines which directly use a model to generate predictions, and also support logprobs (`HFAutoModelInferenceEngine` and `HFLlavaInferenceEngine`), but I didn't come up with some better way to do so.

Maybe just remove the `HFInferenceEngineBase` class, and keep separate implementations for engines:

`HFPipelineBasedInferenceEngine` -> `HFPeftInferenceEngine`

`HFInferenceEngine` (maybe some better name could be used for this) -> `HFAutoModelInferenceEngine` && `HFLlavaInferenceEngine`

What do you think?